be/src/olap/memtable_flush_executor.h

```diff
@@ -38,132 +32,65 @@ class DeltaWriter;
-            _last_flush_status(OLAP_SUCCESS),
-            _counter_cond(0),
-            _flush_executor(flush_executor),
-            _is_cancelled(false) {
-    }
-
-    // submit a memtable to flush. return error if some previous submitted MemTable has failed
-    OLAPStatus submit(std::shared_ptr<MemTable> memtable);
-    // wait for all memtables submitted by itself to be finished.
+    explicit FlushToken(std::unique_ptr<ThreadPoolToken> flush_pool_token)
+        : _flush_status(OLAP_SUCCESS),
+          _flush_token(std::move(flush_pool_token)) {}
+
+    OLAPStatus submit(std::shared_ptr<MemTable> mem_table);
+
+    // error has happpens, so we cancel this token
+    // And remove all tasks in the queue.
+    void cancel();
+
+    // wait all tasks in token to be completed.
     OLAPStatus wait();
+
     // get flush operations' statistics
     const FlushStatistic& get_stats() const { return _stats; }
 
-    bool is_cancelled() {
-        return _last_flush_status.load() != OLAP_SUCCESS || _is_cancelled.load();
-    }
-    void cancel() { _is_cancelled.store(true); }
-
-    // These on_xxx() methods are callback when flush finishes or cancels, user should
-    // not call them directly.
-    // called when a memtable is finished by executor.
-    void on_flush_finished(const FlushResult& res);
-    // called when a flush memtable execution is cancelled
-    void on_flush_cancelled() {
-        _counter_cond.dec();
-    }
-
 private:
-    // flush queue idx in memtable flush executor
-    int32_t _flush_queue_idx;
-    // the flush status of last memtable
-    std::atomic<OLAPStatus> _last_flush_status;
-    // used to wait/notify the memtable flush execution
-    CounterCondVariable _counter_cond;
+    void _flush_memtable(std::shared_ptr<MemTable> mem_table);
 
+    OLAPStatus _flush_status;
+    std::unique_ptr<ThreadPoolToken> _flush_token;
     FlushStatistic _stats;
-    MemTableFlushExecutor* _flush_executor;
-
-    // the caller of the flush handler can set this variable to notify that the
-    // uppper application is already cancelled.
-    std::atomic<bool> _is_cancelled;
 };
 
 // MemTableFlushExecutor is responsible for flushing memtables to disk.
-// Each data directory has a specified number of worker threads and each thread will correspond
-// to a queue. The only job of each worker thread is to take memtable from its corresponding
-// flush queue and writes the data to disk.
-//
-// NOTE: User SHOULD NOT call method of this class directly, use pattern should be:
+// It encapsulate a ThreadPool to handle all tasks.
+// Usage Example:
 //      ...
 //      std::shared_ptr<FlushHandler> flush_handler;
-//      memTableFlushExecutor.create_flush_handler(path_hash, &flush_handler);
+//      memTableFlushExecutor.create_flush_token(path_hash, &flush_handler);
 //      ...
-//      flush_handler->submit(memtable)
+//      flush_token->submit(memtable)
 //      ...
 class MemTableFlushExecutor {
```

Comment:
The only class the user needs to know is `FlushToken`.
So maybe we should hide the implementation of `MemTableFlushExecutor` in `.cpp` file, because the user does not need to directly know it.  But this can be refactored later.